dual_net.py

```diff
@@ -321,58 +193,133 @@ def my_res_layer(inputs, train_mode):
+    # Register metrics as tf.summary to make them show up during training
+    for metric_name, metric_op in metric_ops.items():
+        tf.summary.scalar(metric_name, metric_op[1])
+
+    return tf.estimator.EstimatorSpec(
+        mode=mode,
+        predictions={
+            'policy_output': policy_output,
+            'value_output': value_output,
+        },
+        loss=combined_cost,
+        train_op=train_op,
+        eval_metric_ops=metric_ops,
+        )
+
+
+def get_estimator(working_dir, **hparams):
+    hparams = get_default_hyperparams(**hparams)
+    return tf.estimator.Estimator(
+        model_fn,
+        model_dir=working_dir,
+        params=hparams)
+
+
+def bootstrap(working_dir, **hparams):
+    """Initialize a tf.Estimator run with random initial weights.
+
+    Args:
+        working_dir: The directory where tf.estimator will drop logs,
+            checkpoints, and so on
+        hparams: hyperparams of the model.
+    """
+    hparams = get_default_hyperparams(**hparams)
+    # a bit hacky - forge an initial checkpoint with the name that subsequent
+    # Estimator runs will expect to find.
+    #
+    # Estimator will do this automatically when you call train(), but calling
+    # train() requires data, and I didn't feel like creating training data in
+    # order to run the full train pipeline for 1 step.
+    estimator_initial_checkpoint_name = 'model.ckpt-1'
+    save_file = os.path.join(working_dir, estimator_initial_checkpoint_name)
+    sess = tf.Session(graph=tf.Graph())
+    with sess.graph.as_default():
+        features, labels = get_inference_input()
+        model_fn(features, labels, tf.estimator.ModeKeys.PREDICT, hparams, None)
+        sess.run(tf.global_variables_initializer())
+        tf.train.Saver().save(sess, save_file)
+
+
+def export_model(working_dir, model_path):
+    """Take the latest checkpoint and export it to model_path for selfplay.
+
+    Assumes that all relevant model files are prefixed by the same name.
+    (For example, foo.index, foo.meta and foo.data-00000-of-00001).
+
+    Args:
+        working_dir: The directory where tf.estimator keeps its checkpoints
+        model_path: The path (can be a gs:// path) to export model to
+    """
+    estimator = tf.estimator.Estimator(model_fn, model_dir=working_dir,
+        params='ignored')
+    latest_checkpoint = estimator.latest_checkpoint()
+    all_checkpoint_files = tf.gfile.Glob(latest_checkpoint + '*')
+    for filename in all_checkpoint_files:
+        suffix = filename.partition(latest_checkpoint)[2]
+        destination_path = model_path + suffix
+        print("Copying {} to {}".format(filename, destination_path))
+        tf.gfile.Copy(filename, destination_path)
+
+
+def train(working_dir, tf_records, generation_num, **hparams):
+    assert generation_num > 0, "Model 0 is random weights"
+    estimator = get_estimator(working_dir, **hparams)
+    max_steps = generation_num * EXAMPLES_PER_GENERATION // TRAIN_BATCH_SIZE
```

Comment:
this makes TRAIN_BATCH_SIZE immutable for the length of a run, correct?  I.e., if i lowered it on the fly, the next training run would be WAY more steps, correct?  Can this be based off the last checkpoint global_step instead?

Reply:
Yes, this would make the batch size and examples_per_generation immutable per run.

It'd make the code a bit uglier to pull the global_step in from the previous checkpoint, but I agree it'd make for a more robust calculation.

Reply:
yeah, sorry :(  In theory, they shouldn't change for a run but in practice they might;